backend/backend.py

```diff
@@ -0,0 +1,182 @@
+
+        self._on_loop()
+
+        if time() > self.prev_recording_status_time + self.recording_status_interval:
+            self.send_recording_status()
+
+    @abstractmethod
+    def _on_loop(self) -> None:
+        """ Implemented by overlays to perform any neccessary operations that
+            should be performed on a regular period. This may involve updating
+            the display, which may be blocking. Should not be called outside
+            of the Backend class. """
+
+    @abstractmethod
+    def stop_video(self) -> None:
+        """ Stops displaying the video feed and releases any resources
+            captured. """
+
+    def start_recording(self) -> None:
+        """ Starts an h264 recording with the first available name located in
+            the recordings folder.
+
+            Should be paired with a call to stop_recording. Does nothing if
+            already recording. """
+
+        # If we're already recording just respond with a status
+        if self.recording:
+            self.send_recording_status()
+            return
+
+        # Ensure output folder exists
+        output_folder = Path(__file__).parent.parent / "recordings"
+        output_folder.mkdir(exist_ok=True)
+
+        # Find next available video filename
+        video_number = 1
+        output_file_pattern = "rec_{}.h264"
+        while (output_folder / output_file_pattern.format(video_number)).exists():
+            video_number += 1
+        self.recording_output_file = str(output_folder / output_file_pattern.format(video_number))
+
+        # Start recording. Send an error if one occurs, otherwise a status.
+        try:
+            self._start_recording()
+        except Exception:
+            self.recording = False
+            self.send_recording_error()
+        else:
+            self.send_recording_status()
+
+    def _start_recording(self) -> None:
+        """ Starts recording to the file at self.recording_output_file. Should
+            not be called outside of the Backend class. """
+
+        raise NotImplementedError(f"Recording is not supported with {type(self).__name__}")
+
+    def stop_recording(self) -> None:
+        """ Stops and saves any current recording at the location found in
+            self.recording_output_file.
+
+            Should be paired with a call to start_recording. No action is taken
+            if there was no recording in progress. """
+
+        self.recording = False
+        try:
+            self._stop_recording()
+        except Exception:
+            self.send_recording_error()
+        else:
+            self.send_recording_status()
+
+    def _stop_recording(self) -> None:
+        """ Stops recording and saves to the file at self.recording_output_file.
+            Should not be called outside of the Backend class. """
+
+        raise NotImplementedError(f"Recording is not supported with {type(self).__name__}")
+
+    def check_recording_errors(self) -> None:
```

Comment:
Is this an abstract method? Or is leaving this blank intended?

Reply:
Good question. I didn't want to make it abstract because for the time being, I think it's okay that backends are not required to support recording. (OpenCV doesn't right now, and it wouldn't be super useful as we only use OpenCV for testing on our laptops)